### ISSUE TITLE
Adding watch error: "too many open files"

### ISSUE THREAD

**Issue by probabble** [GoogleContainerTools/skaffold#272]
Issues without logs and details are more complicated to fix.
Please help us!

### Expected behavior

Skaffold runs and re-runs on changes to watched files

### Actual behavior

Skaffold runs and then fails when adding watches. My project is quite large

### Information
```ERRO[0101] run: running skaffold steps: running watch: adding watching for deps: starting watch on file /Users/evan/PycharmProjects/portal/portal/apps/people/api/urls.py: adding watch for /Users/evan/PycharmProjects/portal/portal/apps/people/api/urls.py: too many open files ```

```
Skaffold version: 0.2
Operating system: osx 10.13.3
# skaffold.yaml
apiVersion: skaffold/v1alpha1
kind: Config
build:
  artifacts:
  - imageName: portal
    workspace: .
  local: {}
deploy:
  kubectl:
    manifests:
    - paths:
      - docker/kubernetes-portal.yaml
      parameters:
        IMAGE_NAME: portal
```

### Steps to reproduce the behavior

  1. in a large project, run `skaffold run -f skaffold.yml` (this works)
  2. run `skaffold dev` (this breaks)

**Comment by r2d4** [comment]
Can you tell us how many files you have in your directory? I've also hit some limits on different OSes.

**Comment by probabble** [comment]
I'm at 55,557 files ... so, kind of a lot!

**Comment by r2d4** [comment]
Do you mind sharing your dockerfiles? As a workaround for now, if you can refactor your dockerfiles so you don't have such a large dependency graph, you'll be able to have quicker build and deploy cycles and skaffold won't fail.

We currently respect the .dockerignore file also and are planning some sort of skaffold ignore UX

**Comment by probabble** [comment]
Hi @r2d4 -- I don't think my `.dockerignore` file is getting respected by the file watcher... I see my compiled python files getting added to the watcher despite the first line in my .dockerignore file being ```*.pyc``` ... so that might help reduce the number of files.

I upgraded to 0.3 today and still see the same behavior.

**Comment by probabble** [comment]
but it looks like proper support for .dockerignore is coming soon: https://github.com/GoogleCloudPlatform/skaffold/pull/338

**Comment by antonversal** [comment]
I have same issue with v0.4.0. I have ~ 4k files. Is it a lot? 

my workaround is to set `ulimit -S -n 4048`

**Comment by antonversal** [comment]
I run ` skaffold dev -v debug`. What I can see is `Recursively adding data` for all excluded folders in `.dockerignore` and then ignoring of each file. Because of that startup time is super long and probably it causes the issue `too many open files`
log(short version):
```
DEBU[0000] Recursively adding data                      
INFO[0000] deps [data/node_modules/jest-config/build/defaults.js data/node_modules/join-protos/node_modules/grpc/deps/grpc/src/core/ext/transport/chttp2/transport/frame_settings.cc data/node_modules/@types/strip-json-comments/README.md data/node_modules/node-pre-gyp/lib/util/nw-pre-gyp/index.html data/node_modules/sshpk/lib/errors.js data/node_modules/core-js/modules/es6.string.includes.js data/node_modules/fsevents/node_modules/mime-db/index.js
...
...
DEBU[0000] Skipping excluded path: data/node_modules/jest-config/build/defaults.js 
DEBU[0000] Skipping excluded path: data/node_modules/join-protos/node_modules/grpc/deps/grpc/src/core/ext/transport/chttp2/transport/frame_settings.cc 
DEBU[0000] Skipping excluded path: data/node_modules/@types/strip-json-comments/README.md 
```

**Comment by jonbcampos** [comment]
+1 saw this in a node application. the node_modules/ file ends up with a lot of files being watched

**Comment by berstend** [comment]
Ran into the same issue today, I'm using `ulimit -S -n 9999 && skaffold dev` as a temporary solution for now.

**Comment by wiemann** [comment]
make that `ulimit -S -n 30000` for SPA frameworks like VueJS with a couple of plugins added.

**Comment by mailaneel** [comment]
this is happening to me as well, when I have `COPY . .` in my Dockerfile
and `.dockerignore` has `node_modules`

right now I am only copying folders/files required instead

**Comment by ajbouh** [comment]
I implemented a new watcher based on Facebook's out of process watchman service. Under the hood it's using fsevents (on macoOS) which does not appear to have the same file descriptor issues that kevents does.

Initial results are promising. What's the interest level for making a PR?

**Comment by dlorenc** [comment]
We have an mtime watcher that shouldn't have this problem.

**Comment by DHBuild002** [comment]
Hi Guys, After a lot of twoing and frowing I was able to fix this problem on my own personal project.

The issue it turned out for me atleast was due to an old version of babel-core that I had installed. 

I removed that from my JSON file, then uninstalled it, and the next test I ran worked succesfully.

It did not complain there were two many files running. I think it was because I had conflicting copies of Babel running at the same time that was the issue.

Example:
  "devDependencies": {
    "babel-core": "6.2.1" 
    "@babel/preset-env": "^7.5.5",
    "jest": "^24.8.0",
    "parcel-bundler": "^1.12.3"
  },

Hope this helps!
